File: src/video_processing/trimVideo.m
function outputPath = trimVideo(videoPath, startTime, endTime, outputPath)
% ============================================================
% trimVideo
% ============================================================
%
% PURPOSE:
% This function extracts only the required portion of a video.
%
% INPUTS:
%
% videoPath  -> Full path of the original video
%
% startTime  -> Time (in seconds) where the required video starts
%
% endTime    -> Time (in seconds) where the required video ends
%
% outputPath -> Where the trimmed video should be saved
%
% OUTPUT:
%
% outputPath -> Path of the newly created trimmed video
%
% ============================================================


%% 1. Open the original video

% VideoReader allows us to read the original video
% frame-by-frame.

video = VideoReader(videoPath);


%% 2. Validate the start and end times

% Start time cannot be negative.

if startTime < 0
    error("Start time cannot be negative.");
end


% End time must be greater than start time.

if endTime <= startTime
    error("End time must be greater than start time.");
end


% End time cannot be beyond the actual video duration.

if endTime > video.Duration
    error("End time exceeds the video duration.");
end


%% 3. Create a VideoWriter object

% VideoWriter allows us to create a new video file.

outputVideo = VideoWriter(outputPath, "Motion JPEG AVI");


% Keep the same frame rate as the original video.

outputVideo.FrameRate = video.FrameRate;


% Open the output video for writing.

open(outputVideo);


%% 4. Move to the beginning of the required section

% CurrentTime tells VideoReader where we currently are
% in the video.

video.CurrentTime = startTime;


%% 5. Read only the required section

while hasFrame(video)

    % Stop when we reach the requested end time.

    if video.CurrentTime > endTime
        break;
    end


    % Read the current frame.

    frame = readFrame(video);


    % Write the frame into the new video.

    writeVideo(outputVideo, frame);

end


%% 6. Close the output video

% Always close VideoWriter after writing.

close(outputVideo);


%% 7. Tell the user where the file was created

fprintf("\nTrimmed video created successfully.\n");
fprintf("Saved at: %s\n", outputPath);

end

File: src/video_processing/trim_video.py
import cv2


def trim_video(video_path, start_time, end_time, output_path):
    """
    Trim a video and save only the requested portion.

    Parameters
    ----------
    video_path : str
        Path to the original video.

    start_time : float
        Start time of the required portion, in seconds.

    end_time : float
        End time of the required portion, in seconds.

    output_path : str
        Path where the trimmed video should be saved.
    """

    # ============================================================
    # 1. Open the original video
    # ============================================================

    video = cv2.VideoCapture(video_path)

    if not video.isOpened():
        raise ValueError(f"Could not open video: {video_path}")


    # ============================================================
    # 2. Get video information
    # ============================================================

    fps = video.get(cv2.CAP_PROP_FPS)
    total_frames = int(video.get(cv2.CAP_PROP_FRAME_COUNT))

    duration = total_frames / fps


    # ============================================================
    # 3. Validate the start and end times
    # ============================================================

    # Start time cannot be negative.

    if start_time < 0:
        video.release()
        raise ValueError("Start time cannot be negative.")


    # End time must be greater than start time.

    if end_time <= start_time:
        video.release()
        raise ValueError("End time must be greater than start time.")


    # End time cannot exceed video duration.

    if end_time > duration:
        video.release()
        raise ValueError(
            f"End time exceeds the video duration ({duration:.2f} seconds)."
        )


    # ============================================================
    # 4. Create the output video
    # ============================================================

    width = int(video.get(cv2.CAP_PROP_FRAME_WIDTH))
    height = int(video.get(cv2.CAP_PROP_FRAME_HEIGHT))

    # MP4 output
    fourcc = cv2.VideoWriter_fourcc(*"mp4v")

    output_video = cv2.VideoWriter(
        output_path,
        fourcc,
        fps,
        (width, height)
    )

    if not output_video.isOpened():
        video.release()
        raise ValueError(f"Could not create output video: {output_path}")


    # ============================================================
    # 5. Move to the starting position
    # ============================================================

    video.set(cv2.CAP_PROP_POS_MSEC, start_time * 1000)


    # ============================================================
    # 6. Read and write only the required section
    # ============================================================

    while True:

        ret, frame = video.read()

        if not ret:
            break

        # Current position in milliseconds
        current_time = video.get(cv2.CAP_PROP_POS_MSEC) / 1000

        # Stop when we reach the requested end time
        if current_time > end_time:
            break

        output_video.write(frame)


    # ============================================================
    # 7. Close the videos
    # ============================================================

    video.release()
    output_video.release()


    # ============================================================
    # 8. Tell the user where the file was created
    # ============================================================

    print("\nTrimmed video created successfully.")
    print(f"Saved at: {output_path}")